File: generated_code/python/solution_3813_2.py
from typing import List, Dict
from collections import Counter
import math

class Solution:
    def smallestPalindrome(self, s: str, k: int) -> str:
        # Step 1: Count character frequencies
        counts = Counter(s)
        
        # Step 2: Validate if a palindrome can be formed
        # At most one character can have an odd count
        odd_chars = [char for char, count in counts.items() if count % 2 != 0]
        if len(odd_chars) > 1:
            return ""
        
        # Determine the middle character if it exists
        mid_char = odd_chars[0] if odd_chars else ""
        
        # Step 3: Construct the pool of characters for the first half
        # Sorting ensures we can build the palindrome lexicographically
        half_chars = []
        for char in sorted(counts.keys()):
            half_chars.extend([char] * (counts[char] // 2))
            
        n = len(half_chars)
        
        # Precompute frequencies of characters in the first half to calculate permutations
        half_counts = Counter(half_chars)
        
        # Helper to compute total unique permutations for a given remaining frequency map
        def get_total_permutations(f_map: Dict[str, int], total_len: int) -> int:
            if total_len == 0:
                return 1
            ans = math.factorial(total_len)
            for count in f_map.values():
                if count > 1:
                    ans //= math.factorial(count)
            return ans

        # Step 4: Digit-by-digit (or character-by-character) construction of the first half
        # We try to place the smallest available character at the current position.
        # If the number of permutations possible with that choice is less than k, 
        # we skip all those permutations, subtract the count from k, and try the next character.
        first_half = []
        remaining_len = n
        
        # Track unique sorted keys available for correct ordering during iteration
        unique_sorted_chars = sorted(half_counts.keys())
        
        for i in range(n):
            placed = False
            for char in unique_sorted_chars:
                if half_counts[char] == 0:
                    continue
                
                # Try placing 'char' at index i
                half_counts[char] -= 1
                
                # Calculate how many permutations can be formed with the remaining characters
                perms = get_total_permutations(half_counts, remaining_len - 1)
                
                if k <= perms:
                    # The k-th permutation lies within this character's branch
                    first_half.append(char)
                    remaining_len -= 1
                    placed = True
                    break
                else:
                    # Skip all permutations starting with this character
                    k -= perms
                    half_counts[char] += 1  # Backtrack
            
            # If we couldn't place any character, k is out of bounds
            if not placed:
                return ""
        
        # If after placing all characters, k is still greater than 1, then k is out of bounds
        if k > 1:
            return ""
            
        # Step 5: Reconstruct the full palindrome from the first half
        first_half_str = "".join(first_half)
        return first_half_str + mid_char + first_half_str[::-1]

File: generated_code/cpp/solution_3813_2.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <numeric>

using namespace std;

class Solution {
private:
    // Helper function to calculate combinations (n Choose r) safely up to 2*10^9
    long long nCr(int n, int r) {
        if (r < 0 || r > n) return 0;
        if (r == 0 || r == n) return 1;
        if (r > n / 2) r = n - r; // Optimize calculations
        
        long long res = 1;
        for (int i = 1; i <= r; ++i) {
            // Prevent potential overflows; checking bounds if necessary, 
            // though n will be small enough since max string length is typically small.
            res = res * (n - i + 1) / i;
        }
        return res;
    }

    // Computes the total number of unique permutations of a multinomial coefficient
    // total_permutations = (sum(counts)!) / (counts[0]! * counts[1]! * ...)
    long long get_permutations_count(const vector<int>& counts) {
        int total_elements = 0;
        for (int c : counts) total_elements += c;
        
        long long total_perms = 1;
        int current_n = total_elements;
        
        for (int c : counts) {
            if (c > 0) {
                total_perms *= nCr(current_n, c);
                current_n -= c;
            }
        }
        return total_perms;
    }

public:
    string smallestPalindrome(string s, int k) {
        vector<int> freq(26, 0);
        for (char c : s) {
            freq[c - 'a']++;
        }

        // Validate if a palindrome can be formed
        int odd_count = 0;
        char odd_char = 0;
        vector<int> half_freq(26, 0);
        int half_len = 0;

        for (int i = 0; i < 26; ++i) {
            if (freq[i] % 2 != 0) {
                odd_count++;
                odd_char = 'a' + i;
            }
            half_freq[i] = freq[i] / 2;
            half_len += half_freq[i];
        }

        // A palindrome can have at most one character with an odd frequency
        if (odd_count > 1) {
            return "";
        }

        // Check if the total possible palindromic permutations is less than k
        long long total_palindromes = get_permutations_count(half_freq);
        if (total_palindromes < k) {
            return "";
        }

        // Construct the first half of the k-th lexicographical palindrome using a digit-by-digit/character-by-character approach
        string first_half = "";
        long long target = k;

        for (int i = 0; i < half_len; ++i) {
            for (int j = 0; j < 26; ++j) {
                if (half_freq[j] > 0) {
                    // Try placing character 'a' + j at the current position
                    half_freq[j]--;
                    long long perms_with_this_char = get_permutations_count(half_freq);

                    if (target <= perms_with_this_char) {
                        // The k-th permutation lies within this block
                        first_half += (char)('a' + j);
                        break; // Move to the next position in the string
                    } else {
                        // The k-th permutation is further ahead, skip this block
                        target -= perms_with_this_char;
                        half_freq[j]++; // Backtrack and try the next character
                    }
                }
            }
        }

        // Form the complete palindrome using the first half, the odd character (if any), and the reversed first half
        string second_half = first_half;
        reverse(second_half.begin(), second_half.end());

        if (odd_count == 1) {
            return first_half + odd_char + second_half;
        } else {
            return first_half + second_half;
        }
    }
};